### PR TITLE
Add RAFT llamapack

### PR DESCRIPTION
**Pull request by ravi03071991**
# Description

PR for creating dataset based on RAFT paper as llamapack.

Fixes # (issue)

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-packs/llama-index-packs-raft-dataset/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+This LlamaPack implements the Corrective Retrieval Augmented Generation (CRAG) [paper](https://arxiv.org/pdf/2401.15884.pdf)
+
+Corrective Retrieval Augmented Generation (CRAG) is a method designed to enhance the robustness of language model generation by evaluating and augmenting the relevance of retrieved documents through a an evaluator and large-scale web searches, ensuring more accurate and reliable information is used in generation.
+
+This LlamaPack uses [Tavily AI](https://app.tavily.com/home) API for web-searches. So, we recommend you to get the api-key before proceeding further.
+
+### Installation
+
+```bash
+pip install llama-index llama-index-tools-tavily-research
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-raft-dataset/README.md:12]: wrong pip install?

File: llama-index-packs/llama-index-packs-raft-dataset/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# Corrective Retrieval Augmented Generation Llama Pack
+
+This LlamaPack implements the Corrective Retrieval Augmented Generation (CRAG) [paper](https://arxiv.org/pdf/2401.15884.pdf)
+
+Corrective Retrieval Augmented Generation (CRAG) is a method designed to enhance the robustness of language model generation by evaluating and augmenting the relevance of retrieved documents through a an evaluator and large-scale web searches, ensuring more accurate and reliable information is used in generation.
+
+This LlamaPack uses [Tavily AI](https://app.tavily.com/home) API for web-searches. So, we recommend you to get the api-key before proceeding further.
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-raft-dataset/README.md:7]: Is this true?

File: llama-index-packs/llama-index-packs-raft-dataset/README.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# Corrective Retrieval Augmented Generation Llama Pack
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-raft-dataset/README.md:1]: I think this is not the correct title or intro actually

File: llama-index-packs/llama-index-packs-raft-dataset/README.md
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-raft-dataset/README.md:1]: I think this readme is not updated

File: llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+from llama_index.core.llama_pack.base import BaseLlamaPack
+from llama_index.core import SimpleDirectoryReader
+
+# from llama_index.core.node_parser import SemanticSplitterNodeParser
+from llama_index.core.llms import ChatMessage
+from llama_index.llms.openai import OpenAI
+from llama_index.core.node_parser import SimpleNodeParser
+
+DEFAULT_CHUNK_SIZE = 512
+# DEFAULT_BREAKPOINT_PERCENTILE_THRESHOLD = 95 # Use when SemanticSplitterNodeParser
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py:18]: There is some commented out code, should we remove it?

File: llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+import random
+from datasets import Dataset
+
+from llama_index.core.llama_pack.base import BaseLlamaPack
+from llama_index.core import SimpleDirectoryReader
+
+# from llama_index.core.node_parser import SemanticSplitterNodeParser
+from llama_index.core.llms import ChatMessage
+from llama_index.llms.openai import OpenAI
+from llama_index.core.node_parser import SimpleNodeParser
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py:15]: SimpleNodeParser is an alias for SentenceSplitter, lets use that instead

File: llama-index-packs/llama-index-packs-raft-dataset/README.md (hunk 1 above)
Comment (ravi03071991) [reply llama-index-packs/llama-index-packs-raft-dataset/README.md:12]: Oh I think it got mixed up with CRAG. Will clean it up.

File: llama-index-packs/llama-index-packs-raft-dataset/README.md
Comment (ravi03071991) [reply llama-index-packs/llama-index-packs-raft-dataset/README.md:1]: Yeah. Will update it.

File: llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py (hunk 4 above)
Comment (ravi03071991) [reply llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py:18]: Yeah, I actually intend to use `SemanticSplitterNodeParser` but I think there seems some pydantic issue with it. So for now shifted to simplenodeparser.

File: llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+DEFAULT_CHUNK_SIZE = 512
+# DEFAULT_BREAKPOINT_PERCENTILE_THRESHOLD = 95 # Use when SemanticSplitterNodeParser
+
+
+class RAFTDatasetPack(BaseLlamaPack):
```
Comment (hatianzhang) [inline llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py:21]: it seems only getting dataset as from https://github.com/ShishirPatil/gorilla/pull/218/files
do we also have fine tuning ?

Comment (ravi03071991) [reply llama-index-packs/llama-index-packs-raft-dataset/llama_index/packs/raft_dataset/base.py:21]: Yeah. Once, we have the dataset we can use HF/ Unsloth/ any other framework for finetuning. They have used Azure platform for finetuning.

**Review by hatianzhang — APPROVED** [review]

**Pull request merged**